### ISSUE TITLE
Dogfooding the CLI

### ISSUE THREAD

**Issue by cloudhead** [radicle-dev/heartwood#148]
I've identified a set of missing functionality in the CLI that once implemented would allow us to dogfood.

*  [x] `rad track <key> --alias <name>` (#151)
    *  [x] DB storage of trackings (#146)
*  [x] `rad assign <issue> @<alias>` (#159)
*  [x] `rad issue` lists assigned issues (#171)
*  [x] `rad issue show <issue>` shows issue (#171)
*  [x] `rad patch open` from any branch (@slack-coder)
*  [x] `rad patch show <id>` shows patch diff (#226)
*  [x] `rad patch checkout <id>` should checkout a new branch with the patch changeset (#304)
*  [x] `rad patch update` when we're not the patch author (@slack-coder )
*  [ ] `rad merge <patch>` should deal with concurrent revisions (@slack-coder )

**Comment by slack-coder** [comment]
:eye: rad assign <issue> @<alias>

**Comment by cloudhead** [comment]
Here's the workflow we're aiming for:


**@cloudhead**

This document assumes a user profile has been setup with `rad auth`.



### Step M.0: Add collaborators

First, we add a collaborator called `dave`. Any name can be chosen
here, since it is simply an alias to make the key more recognizable.
The CLI will from this point on allow us to refer to this key using `@dave`.

    $ rad track z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi --alias dave

We do this for two reasons:

1. It establishes a *trust* relationship between us and @dave, meaning that
we will see social objects published by @dave.
2. It tells our radicle node to synchronize @dave's code in the context of
the current project, so that we have it locally.

Note that `@z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi` is also an alias
for itself. We can use that form interchangeably with `@dave`.

### Step M.1: Open an issue (optional)

Second, we open an issue, giving it a title. The issue CID is displayed.

    $ rad issue open --title "Add install instructions"
    ok: !e45aa5699c0fb3ff5c9b74fd756da8b3880f4927

We assign the issue to @dave. Alternatively, we could also specify his public
key.

    $ rad assign !e45aa56 @dave
    ok: issue e45aa56 assigned to z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi

To do the above in a single command, we could also use the `--assign <user>`
option to `open`.

---

**@dave**

### Step C.0: List assigned issues

@dave lists open issues. The first issue shown is assigned to him, hence the
`@` symbol prefixing the issue.

    $ rad issue
    @ e45aa56  Add install instructions
    ! 93c5eb3  Fix copyright notice

He then displays more information about the issue:

    $ rad issue show e45aa56

    Add install instructions

    Opened by @cloudhead (z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK)

    To help first time users, we should add install instructions to the
    README, under an 'Installation' section.

### Step C.1: Start branch and commit some code

To start working on the issue, he creates a new branch:

    $ git checkout -b dave/issues/e45aa56

And commits some code:

    $ vim
    [...]

    $ git commit -a
    [dave/issues/e45aa56 c5defe9] Add install instructions
    Date: Wed Feb 2 16:56:37 2022 +0100
    1 file changed, 78 insertions(+)

### Step C.2: Open patch

Once the code is ready, @dave opens patch with the changes. The patch CID
is displayed.

    $ rad patch open --closes !e45aa56
    ok: master <- dave/issues/e45aa56 (c5defe9)
    ok: #6c13d86b278f2e08ae5f67e1563164884d764ef9

This will sync:

1. The patch object
2. The branch and commit which includes the change

### Step C.2: Amend patch (Optional)

TODO

---

**@cloudhead**

### Step M.2: Check open patches

We check for open patches in our project:

    $ rad patch
    6c13d86 R1 (!e45aa56)  Add install instructions  @dave

We see a patch with CID `6c13d86` opened by @dave. We inspect the commits:

    $ rad patch show %6c13d86

    patch 6c13d86b278f2e08ae5f67e1563164884d764ef9

        Add install instructions

        * @dave opened 1 day ago

        Closes !e45aa56

    commit c5defe98b8507049e5dc0e0b3a189bcf3f73844f (HEAD -> master)
    Author: Dave <dave@radicle.xyz>
    Date:   Wed Feb 2 16:56:37 2022 +0100

        Add install instructions

    diff --git a/README.md b/README.md
    index 548ab9c..cc8ab86 100644
    --- a/README.md
    +++ b/README.md
    @@ -1,3 +1,10 @@
     # Radicle Client Tools

     > ✨ Command-line client tooling for Radicle.
    +
    +## Installation
    +
    +You can install the radicle tools by running the following command from inside
    +this repository:
    +
    +    cargo install --force --locked --path .

We see a single commit, with the diff.

### Step M.3: Rebase patch (Optional)

In some cases, it may be necessary to re-base the changes onto our branch, if
the histories have diverged.

    $ rad patch checkout %6c13d86
    ok: branch patch/6c13d86 created

    $ git rebase master
    Successfully rebased and updated refs/heads/patch/6c13d86.

    $ rad patch update
    ok: R1 -> R2

The above can also be accomplished with the `rebase` subcommand of `rad-patch`.

### Step M.4: Accept and merge

Since we're happy with the changes, we accept the patch and merge it into
our tree:



    $ rad accept %6c13d86 --merge
    ok: master 548ab9c..cc8ab86
    ok: Deleted branch patch/6c13d86 (was cc8ab86).

This does several things:

1. It publishes a review for the patch, with `{ verdict: "accept" }`.
2. It merges the patch code into our working copy's `master` branch.
3. It pushes our updated `master` to our `rad` remote.
4. It publishes a merge confirmation for the patch, so that others know we merged it.
5. It announces our new `master` to the network.
6. Finally, it deletes the local patch checkout branch

If we display the patch listing now, we'll see some of this information:

    $ rad patch -v

    patch 6c13d86b278f2e08ae5f67e1563164884d764ef9

    Add install instructions

    * @dave opened 1 day ago
    ✔ @cloudhead accepted 1 minute ago
    ✔ @cloudhead merged into master 1 minute ago

### Step M.5: Close issue

    $ rad issue close !e45aa56

**Comment by cloudhead** [comment]
What do you mean by "point to"?

**Comment by slack-coder** [comment]
> What do you mean by "point to"?

An alias for, 'rad assign' is an alias for 'rad issue assign' when the Object ID is for an Issue.

**Comment by slack-coder** [comment]
@cloudhead I'm taking on 'rad patch open' and 'rad patch show'

**Comment by slack-coder** [comment]
`rad patch open` is complete except for:
  - the `--closes [issue]` tag and similar arguments mentioned in the issue must be added.
  - CI needs a long term fix for documentations testing commands which use `radicle-remote-helper`.

**Comment by slack-coder** [comment]
`rad::patch::create` should be split into a create or update function as part of these changes.

**Comment by vincenzopalazzo** [comment]
On Wed Jan 18, 2023 at 6:16 PM CET, Slack Coder wrote:
> `rad::patch::create` should be split into a create or update function as part of these changes.

there is any reason to do not use a flag like `rad patch create --update`?

**Comment by slack-coder** [comment]
> there is any reason to do not use a flag like `rad patch create --update`?

The questions can be reversed, why use '--update' over 'update'?  The later is shorter

**Comment by slack-coder** [comment]
@cloudhead what's the use case for `rad patch update` when your not the creator and concurrent `rad merge` with concurrent revisions?

The naive understanding would be to have a single owner of patchs.

**Comment by cloudhead** [comment]
> @cloudhead what's the use case for `rad patch update` when your not the creator and concurrent `rad merge` with concurrent revisions?
> 
It allows collaboration on a patch, simple as that. Or for example, a patch may be abandonned, and instead of creating a new patch and copying the changes over, you can just update it and merge it.

Re: merges, we just have to make sure that we're able to select which revision to merge if there's more than one, and there is no "newest" revision. I'm wondering though if it makes sense to have an explicit parent pointer on revisions, or if the clock suffices. I guess this only makes a difference in a multi-author scenario..

**Comment by slack-coder** [comment]
> I'm wondering though if it makes sense to have an explicit parent pointer on revisions, or if the clock suffices.

Please elaborate, I'm not quite sure what you mean

**Comment by FintanH** [comment]
I think it would be a good idea to add `rad id` to the list above so that we can check:
* Adding a delegate with threshold 1 will correctly verify the document
* Updating the threshold to 2 means the last verified document will be in the current state, and then after exchanging both signatures, the verified document is the new state.

**Comment by cloudhead** [comment]
You mean these are things that aren't done yet, or just that need to be checked/tested?

**Comment by FintanH** [comment]
The first bullet point is tested in the CLI examples. For the second, I think I was still a bit green on writing CLI tests so it hasn't been fully tested.

If we're dogfooding, then perhaps it's an opportunity to add delegates to our own projects.

**Comment by slack-coder** [comment]
@cloudhead I manually tested `rad patch update when we're not the patch author` and its working out of the box.  I'm hoping to include this in a long documentation test with clone, issue, + patch, for a full use example, but I'm blocked by #410.

The priority is now on `rad merge` and I'll return to the documentation test after.

**Comment by cloudhead** [comment]
Great, will have a look at #410

**Comment by slack-coder** [comment]
@cloudhead for 'rad merge <patch> should deal with concurrent revisions ', what do you mean by concurrent revisions?  Would a brief example be possible?

**Comment by cloudhead** [comment]
So imagine a patch has two revisions, and these revisions are created concurrently, by different users. So neither of them came first.

The `merge` command should be able to unambiguously merge either of them. I think this is just a matter of the merge command taking the revision id, which is unique, and not a revision number, which can be duplicate.

**Comment by slack-coder** [comment]
@cloudhead is this now closed?

**Comment by cloudhead** [comment]
Yup, let's close it!
